### PR TITLE
fix: only apply `drop_last` to the classify train loader under `compile`, not val

### PR DESCRIPTION
**Pull request by JESUSROYETH**
`ClassificationTrainer.get_dataloader()` passes `drop_last=self.args.compile` for every mode, so under `compile=True` the validation loader drops samples too, not just the training one. `DetectionTrainer.get_dataloader()` already guards this correctly with `drop_last=self.args.compile and mode == "train"` — both lines come from the same commit (#21975), the guard on the classify side just never got added.

Since `rank=-1` (single GPU) leaves `sampler=None`, and `classify/train.py` never sets `shuffle=` explicitly, `build_dataloader()`'s default `shuffle=True` stays on for validation too. So under `compile=True` the val loader gets `shuffle=True` and `drop_last=True` at the same time, every validation during training sees a smaller, random, different subset of the val set. That corrupts `top1`/`top5` per epoch, the `fitness` that decides `best.pt`, and the patience counter for early stopping.

`final_eval()` doesn't escape it either. `self.test_loader` gets built once in `_setup_train()` (`engine/trainer.py:291`) and reused as-is for every call after that. `self.validator.args.compile = False` at `engine/trainer.py:950` only skips re-compiling the model for the last run (`engine/validator.py:238`) — it never touches `self.dataloader`, and `engine/validator.py:235` only rebuilds that when it's still `None`. So the metrics logged for `best.pt` at the end of training are corrupted the same way as every epoch in between.

## Changes

One line, mirrors the existing `detect/train.py` guard exactly:

```diff
- drop_last=self.args.compile
+ drop_last=self.args.compile and mode == "train"
```

I also checked whether `shuffle` needed the same treatment. `detect/train.py` forces `shuffle=mode=="train"` too, but for the single-process case this repro covers (`rank=-1`, `sampler=None`) it doesn't matter: `shuffle=True` alone, without `drop_last`, doesn't corrupt accuracy, every sample is still seen exactly once, order-independent, in eval mode.

That doesn't hold for multi-GPU though. When `rank != -1`, `shuffle=True` makes `build_dataloader()` pick `torch.utils.data.DistributedSampler(dataset, shuffle=True)` (`data/build.py`). Its default `drop_last=False` pads the index list to a multiple of `world_size` by repeating a few samples whenever `len(dataset) % world_size != 0` — confirmed this locally with a 2-replica/13-item sampler, index `3` lands on both ranks. That's a real problem, but it's pre-existing (same with or without this fix, independent of `compile`/`drop_last`), so it's out of scope for this one-line diff. Leaving it as a separate, unfixed issue instead of claiming `shuffle` is harmless in general.

`BaseTrainer.get_dataloader()` is abstract, every task trainer implements or inherits it. `DetectionTrainer` has the correct guard, and every other trainer (`Segmentation`, `OBB`, `Pose`, `World`, `Depth`, `SemanticSegmentation`, `RTDETR`, the `YOLOE*` variants) inherits from `DetectionTrainer` without overriding `get_dataloader()`. So they're all already covered for the trainer-driven validation path, the one `self.validate()` actually uses during `model.train()`, via `self.test_loader`. `ClassificationTrainer` is the only trainer that extends `BaseTrainer` directly with its own `get_dataloader()` — that's why it's the only place with this bug, and the only place this PR touches.

Scope note: this doesn't cover standalone validation (`model.val(compile=True)` / `yolo val ... compile=True`, no trainer involved). `DetectionValidator.get_dataloader()` (`models/yolo/detect/val.py`) has the same unguarded `drop_last=self.args.compile`. There's no `mode` to gate on there, since standalone validation is always "val", and it's inherited unmodified by 11 descendant validators (`Segmentation`, `OBB`, `Pose`, `World`, `Depth`, `SemanticSegmentation`, `RTDETR`, `NAS`, `YOLOEDetect`, `YOLOESeg`, `FastSAM`). Trainer-driven validation never calls it — `BaseValidator.__call__` only calls `self.get_dataloader()` in the standalone branch, `self.training is False`. So it's a separate bug sharing the same root cause, not something this diff touches. Leaving it for a follow-up PR to keep this one mirrored 1:1 to the existing `detect/train.py` guard.

Deleted: nothing. Net +6/-1 is ruff wrapping the call onto multiple lines once it grows past 120 chars with the new keyword argument, not new logic.

## Validation

Reproduced by calling `ClassificationTrainer.get_dataloader(val_dir, batch_size=7, mode="val")` directly, the exact method this diff changes, against a local 10-class dataset (12 train / 12 val images), Python 3.12, torch 2.10.0+cu128, CPU. Before the fix that call gives `drop_last=True`, 7 of 12 images, and two consecutive passes over the same loader return different label sets. After the fix it gives `drop_last=False`, all 12 images, same set on both passes. The train loader keeps `drop_last=True` under `compile=True` before and after, unchanged.

One caveat: this calls `get_dataloader()` directly with `batch_size=7`, not through `model.train()`. The real entry point doubles the val batch to `batch_size*2=14` (`_build_train_pipeline`, `engine/trainer.py`), and `build_dataloader()` then clamps `batch = min(batch, dataset_len)` to `12` — with this exact 12-image val split, `12 % 12 == 0` and `drop_last` would land on `False` regardless of `compile`. So this small synthetic dataset doesn't reproduce the bug end-to-end through `model.train()`, only by calling the fixed method directly. It still reproduces in real training runs, since real val splits are almost always larger than `batch*2` and don't divide evenly.

`pytest tests/test_engine.py -k "classify or test_task"`: 8 passed. `pytest tests/test_python.py -k "classify or dataloader or Classify"`: 10 passed. Neither suite touches `compile=True` (checked with `rg -n compile` on both files, no matches), so these just show the change doesn't break classify training or dataloader construction — not new coverage of the `drop_last` behaviour itself. That's verified by the direct repro above, not by these two runs.

No new test added.. this mirrors an already-covered PyTorch DataLoader mechanic (`drop_last`/`shuffle`), and neither #21975, which introduced the bug, nor #25162, a related compile/validation fix, added test coverage for this area either.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Fix classification data loading so `drop_last` is enabled only for the training loader when compilation is active.

### 📊 Key Changes
- Passes `drop_last=self.args.compile and mode == "train"` to the classification dataloader.
- Keeps validation batches intact, even when `compile` is enabled.

### 🎯 Purpose & Impact
- Prevents unnecessary dropping of samples from the classification validation set.
- Preserves the intended compiled training behavior while improving validation accuracy and coverage. ✅

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @JESUSROYETH, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This automated message helps with the initial review, and an engineer will assist with any questions. Please review the following checklist:
- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — APPROVED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Clean, targeted fix. The classification trainer now matches the detection trainer by applying `drop_last` only to compiled training loaders, preserving all validation samples under `compile=True`.

**Comment by glenn-jocher** [comment]
The guard matches `DetectionTrainer` and keeps compiled training behavior unchanged. Please add a focused regression test asserting `drop_last=True` for `mode="train"` and `False` for `mode="val"` when `compile=True`; the standalone validation issue can remain a separate follow-up.
